### ISSUE TITLE
Clarification on the Validation Split and Best-Checkpoint Selection for LEVIR-CD-Point

### ISSUE THREAD

**Issue by Ziegher** [Wanderlust717/CARGNet#3]
Hello, thank you for releasing the code and datasets for CARGNet.

I am currently trying to reproduce the experiments reported in the following paper:

> **Point Label Meets Remote Sensing Change Detection: A Consistency-Aligned Regional Growth Network**

During reproduction, I noticed a possible inconsistency between the experimental protocol described in the paper and the released dataset and training code. I would appreciate it if the authors could clarify the following points.

## 1. Dataset Split Described in the Paper

The paper states that the LEVIR-CD-Point dataset contains:

- 445 original image pairs for training
- 64 original image pairs for validation
- 128 original image pairs for testing

After cropping the original images into non-overlapping \(256 \times 256\) patches, the paper reports:

- 7,120 training patches
- 1,024 validation patches
- 2,048 testing patches

The paper also states:

> Validation is performed after every 1000 iterations, and the best model on the validation set is used for evaluation on the test set.

Therefore, according to the paper, the expected experimental protocol appears to be:

1. Train the model on the training set.
2. Evaluate the model on a separate validation set every 1000 iterations.
3. Select the best checkpoint according to validation-set performance.
4. Evaluate the selected checkpoint once on the test set.

## 2. Dataset Structure in the Released Baidu Netdisk Package

I downloaded the LEVIR-CD-Point dataset using the Baidu Netdisk link provided in the repository README.

However, the released dataset appears to contain only `train` and `test` subsets. I could not find a separate `val` or `validation` subset.

In addition, I counted the files in the released `test` image directory and found exactly **2,048 image patches**.

This number matches the test-set size reported in the paper, rather than the reported validation-set size of 1,024 patches.

Therefore, the released `test` directory appears to correspond to the official test split described in the paper, rather than to a validation split that was only named `test` in the code.

## 3. Best-Checkpoint Selection in `train.py`

In the released training code, the model is evaluated every 1000 iterations using the path stored in `args.data_test_path`:

```python
if (i + 1) % 1000 == 0:
    torch.save(
        generator.state_dict(),
        '%s/generator_latest.pth' % args.model_result_dir
    )

    mask_save_path = '%s/test/' % args.model_result_dir
    image_path = args.data_test_path

    test(generator, mask_save_path, image_path)

    gt_path = args.label_path
    mae1, fmeasure1, recall, precision, OA, _ = eval1(
        mask_save_path,
        gt_path,
        2
    )
```

The best checkpoint is then selected according to the F1 score calculated on this dataset:

```python
if fmeasure1[1] > log['best_fm']:
    log['bestfm_iter'] = i
    log['best_fm'] = fmeasure1[1]
    log['mae'] = mae1
    log['R'] = recall[1]
    log['P'] = precision[1]

    torch.save(
        generator.state_dict(),
        '%s/generator_best.pth' % args.model_result_dir
    )
```

Therefore, the checkpoint saved as `generator_best.pth` is selected according to `fmeasure1[1]`, which is calculated using the images from `args.data_test_path` and the ground-truth labels from `args.label_path`.

## 4. Default Dataset Paths

The released configuration defines paths such as:

```python
self.parser.add_argument(
    '--data-test-path',
    type=str,
    default='/mnt/vdf/jyq/CARGNet/datasets/DSIFN/test/',
    help='the txt path of the testing images and labels'
)

self.parser.add_argument(
    '--label-path',
    type=str,
    default='/mnt/vdf/jyq/CARGNet/datasets/DSIFN/test/label/',
    help='the path of test label'
)
```

These default paths explicitly point to the `test` directory and the test labels.

For LEVIR-CD-Point, the corresponding testing script also loads `generator_best.pth` and evaluates it using the LEVIR-CD `test` images and labels:

```python
parser.add_argument(
    '--stict',
    default='/mnt/vdf/jyq/CARGNet/saved_models/LEVIR-CD/generator_best.pth',
    type=str
)

parser.add_argument(
    '--image_path',
    default='/mnt/vdf/jyq/CARGNet/datasets/LEVIR-CD/test',
    type=str
)

parser.add_argument(
    '--label-path',
    type=str,
    default='/mnt/vdf/jyq/CARGNet/datasets/LEVIR-CD/test1/label/'
)
```

The evaluation is then performed as follows:

```python
generator2.load_state_dict(torch.load(args.stict))
generator2.eval()

test(
    generator2,
    args.mask_save_path,
    args.image_path
)

mae1, fmeasure1, recall, precision, OA, IOU = eval1(
    args.mask_save_path,
    args.label_path,
    2
)
```

## 5. My Current Understanding

Based on the currently released dataset and code, the public training procedure appears to be:

1. Train the model on the training set.
2. Every 1000 iterations, generate predictions on the released 2,048-patch test set.
3. Calculate the F1 score using the ground-truth labels of that test set.
4. Save `generator_best.pth` when the test-set F1 score improves.
5. Load `generator_best.pth`.
6. Evaluate and report results on the same test set.

In other words, the released implementation appears to select the best checkpoint according to performance on the official test set, rather than on a separate validation set.

I understand that the test labels are not directly used in the training loss or backpropagation. However, if the test-set F1 score is used to select the best checkpoint, the test set is still involved in model selection, and the final test result is no longer obtained from a fully independent test set.

## Questions

Could the authors please clarify the following questions?

1. Was a separate validation set containing 1,024 patches used to produce the LEVIR-CD-Point results reported in the paper?

2. If a separate validation set was used, could you please provide the validation images, labels, or file lists used for that split?

3. Why is the validation subset described in the paper not included in the released Baidu Netdisk dataset?

4. Are `args.data_test_path` and `args.label_path` in the released training code intended to point to the validation set during training?

5. If these arguments were intended to point to the validation set, what exact paths and files should be used?

6. Was the released `generator_best.pth` selected according to performance on a separate validation set, or according to performance on the released 2,048-patch test set?

7. Were the results in Table III produced using the current public training code, or using a different internal version of the code?

8. Could you please provide the exact dataset split, file lists, training command, configuration, random seed, and code commit used to reproduce the LEVIR-CD-Point results in Table III?

9. If the current public code evaluates the test set during training because of an accidental path or configuration issue, could you please provide a corrected training script or configuration that uses a separate validation set for checkpoint selection?

It is possible that I have misunderstood part of the dataset organization or experimental configuration. If so, please correct my interpretation.

Thank you very much for your time and for making the code and dataset publicly available.
